File: epi/Ruby/06.18.epi.rb
=begin
Compute a random subset of length k of [0,1,2,3,..,n-1]
=end
def random_subset(n, k)
  hash = Hash.new
  0.upto(k-1) do |i|
    ran = rand(i..n-i-1)
    if hash[n-i-1] && hash[ran]
      hash[n-i-1], hash[ran] = hash[ran], hash[n-i-1]
    elsif hash[n-i-1]
      hash[ran] = hash[n-i-1]
      hash[n-i-1] = ran
    elsif hash[ran]
      hash[n-i-1] = hash[ran]
      hash[ran] = n-i-1
    else
      hash[n-i-1] = ran
      hash[ran] = n-i-1
    end
  end
  res = Array.new(k)
  0.upto(k-1) do |i|
    res[i] = hash[n-i-1]
  end
  return res
end

puts random_subset(10, 5).to_s

File: epi/Python/06.18.epi.py
# Compute a random subset
# Design an algorithm that computes an array of size k consisting of distinct
# integers in the set [0..n-1]. All subsets should be equally likely and in
# addition, all permutations of elements of the array should be equally likely.
# Your time complexity should be O(k). Your algorithm can use O(k) space in
# addition to the k element array for the result.
from random import randint
def random_subset(n, k):
    res = []
    hashmap = {}
    for i in xrange(k):
        index = randint(0, n-1-i)
        if index not in hashmap and (n-1-i) not in hashmap:
            res.append(index)
            hashmap[index] = n-1-i
        elif index in hashmap and (n-1-i) not in hashmap:
            res.append(hashmap[index])
            hashmap[index] = n-1-i
        elif index not in hashmap and (n-1-i) in hashmap:
            res.append(index)
            hashmap[index] = hashmap[n-1-i]
        else: # index in hashmap and (n-1-i) in hashmap
            res.append(hashmap[index])
            hashmap[index] = hashmap[n-1-i]
    return res

if __name__ == "__main__":
    print random_subset(10, 5)
